L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc

```diff
@@ -0,0 +1,938 @@
+        if (validTriggerCellFromCells(trigger_cell_id))
+          trigger_cell_det_ids.emplace(trigger_cell_id);
+      }
+    }
+  }
+
+  // HFNose
+  else if (det == DetId::Forward && module_det_id.subdetId() == ForwardSubdetector::HFNose) {
+    HFNoseDetId module_nose_id(module_id);
+    HFNoseDetIdToModule hfn;
+    std::vector<HFNoseTriggerDetId> ids = hfn.getTriggerDetIds(module_nose_id);
+    for (auto const& idx : ids) {
+      if (validTriggerCellFromCells(idx.rawId()))
+        trigger_cell_det_ids.emplace(idx);
+    }
+  }
+  // Silicon
+  else {
+    HGCSiliconDetIdToROC tc2roc;
+    int moduleU = hgc_module_id.moduleU();
+    int moduleV = hgc_module_id.moduleU();
+    unsigned layer = hgc_module_id.layer();
+
+    //Rotate to sector
+    geom_rotation_120_.uvMappingFromSector0(getWaferCentring(layer), moduleU, moduleV, hgc_module_id.sector());
+
+    DetId::Detector det = (hgc_module_id.subdetId() == ForwardSubdetector::HGCEE ? DetId::HGCalEE : DetId::HGCalHSi);
+    HGCalTriggerSubdetector subdet =
+        (hgc_module_id.subdetId() == ForwardSubdetector::HGCEE ? HGCalTriggerSubdetector::HGCalEETrigger
+                                                               : HGCalTriggerSubdetector::HGCalHSiTrigger);
+    unsigned wafer_type = detIdWaferType(det, layer, moduleU, moduleV);
+    int nroc = (wafer_type == HGCSiliconDetId::HGCalFine ? 6 : 3);
+    // Loop on ROCs in wafer
+    for (int roc = 1; roc <= nroc; roc++) {
+      // loop on TCs in ROC
+      auto tc_uvs = tc2roc.getTriggerId(roc, wafer_type);
+      for (const auto& tc_uv : tc_uvs) {
+        HGCalTriggerDetId trigger_cell_id(
+            subdet, hgc_module_id.zside(), wafer_type, layer, moduleU, moduleV, tc_uv.first, tc_uv.second);
+        if (validTriggerCellFromCells(trigger_cell_id.rawId()))
+          trigger_cell_det_ids.emplace(trigger_cell_id);
+      }
+    }
+  }
+
+  return trigger_cell_det_ids;
+}
+
+HGCalTriggerGeometryBase::geom_set HGCalTriggerGeometryV9Imp3::getNeighborsFromTriggerCell(
+    const unsigned trigger_cell_id) const {
+  throw cms::Exception("FeatureNotImplemented") << "Neighbor search is not implemented in HGCalTriggerGeometryV9Imp3";
+}
+
+unsigned HGCalTriggerGeometryV9Imp3::getLinksInModule(const unsigned module_id) const {
+  DetId module_det_id(module_id);
+  unsigned links = 0;
+  // Scintillator
+  if (module_det_id.det() == DetId::HGCalHSc) {
+    links = hSc_links_per_module_;
+  } else if (module_det_id.det() == DetId::Forward && module_det_id.subdetId() == ForwardSubdetector::HFNose) {
+    links = 1;
+  }
+  // TO ADD HFNOSE : getLinksInModule
+  // Silicon
+  else {
+    HGCalDetId module_det_id_si(module_id);
+    unsigned module = module_det_id_si.wafer();
+    unsigned layer = layerWithOffset(module_id);
+    module = (module & sector0_mask_);
+    links = links_per_module_.at(packLayerModuleId(layer, module));
```

Comment:
The packing function used here is not the correct one

Reply:
Changed - I am not sure if `layerWithOffset` still needs to be used here, and if so does the function itself need to be modified?

Reply:
It depends on what is used to fill the `links_per_module_` map. The same packing function should be used here.